src/subs.fs

```diff
@@ -0,0 +1,57 @@
+namespace Elmish
+
+open System
+
+/// SubId - Subscription ID, alias for string
+type SubId = string
+
+/// Subscription - Generates new messages when running
+type Sub<'msg> = { Subscribe: SubId -> Dispatch<'msg> -> IDisposable }
+
+module Sub =
+
+    /// When emitting the message, map to another type
+    let map (f: 'a -> 'msg) (sub: Sub<'a>) : Sub<'msg> =
+        { Subscribe = fun subId dispatch ->
+            sub.Subscribe subId (f >> dispatch) }
+
+module Subs =
+
+    /// When emitting the message, map to another type.
+    /// To avoid ID conflicts with other components, scope SubIds with a prefix.
+    let map (idPrefix: string) (f: 'a -> 'msg) (subs: (SubId * Sub<'a>) list)
+        : (SubId * Sub<'msg>) list =
+        List.map (fun (subId, sub) -> idPrefix + subId, Sub.map f sub) subs
+
+    module Internal =
+
+        let tryStop onError (subId, sub: IDisposable) =
+            try
+                sub.Dispose()
+            with ex ->
+                onError (sprintf "Error stopping subscription: %s" subId, ex)
+
+        let tryStart onError dispatch (subId, newSub: Sub<'msg>) =
+            try
+                Some (subId, newSub.Subscribe subId dispatch)
+            with ex ->
+                onError (sprintf "Error starting subscription: %s" subId, ex)
+                None
+
+        let change onError dispatch (toStop, toKeep, toStart) =
+            toStop |> List.iter (tryStop onError)
+            let started = toStart |> List.choose (tryStart onError dispatch)
+            List.append toKeep started
+
+        let recalc (subs: (SubId * IDisposable) list) (newSubs: (SubId * Sub<'msg>) list) =
+            let keys = subs |> List.map fst |> Set.ofList
+            let newKeys = newSubs |> List.map fst |> Set.ofList
+            if keys = newKeys then
+                [], subs, []
+            else
+                let toKeep, toStop = subs |> List.partition (fun (k, _) -> Set.contains k newKeys)
+                let toStart = newSubs |> List.filter (fun (k, _) -> not (Set.contains k keys))
+                toStop, toKeep, toStart
```

Comment:
Realized this doesn't screen for duplicates. This will lead to incorrect behavior when one of the dupes is turned off but the other isn't. Both will keep running. Preferably needs to error when there is a dupe. Implying "first one wins" behavior. Will fix.

Reply:
Went with last one wins.